### PR TITLE
feat: Add new resource mongodbatlas_cloud_user_project_assignment

### PR DESCRIPTION
**Pull request by csanx**
## Description

Add new resource: `mongodbatlas_cloud_user_project_assignment`

Link to any related issue(s): [CLOUDP-333176](https://jira.mongodb.org/browse/CLOUDP-333176)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/clouduserprojectassignment/resource.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+	newCloudUserProjectAssignmentModel, diags := NewTFModel(ctx, projectID, apiResp)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, newCloudUserProjectAssignmentModel)...)
+}
+
+func fetchProjectUser(ctx context.Context, connV2 *admin.APIClient, projectID, userID, username string) (*admin.GroupUserResponse, error) {
```
Comment (csanx) [inline internal/service/clouduserprojectassignment/resource.go:76]: `fetchProjectUser` function is introduced here and reused in #3569 to extract common functionality between the resource and the data source.

File: internal/service/clouduserprojectassignment/model.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	for _, role := range rolesToRemove {
+		removeRequests = append(removeRequests, &admin.AddOrRemoveGroupRole{
+			GroupRole: role,
+		})
+	}
+
+	return addRequests, removeRequests, nil
+}
+
+func diffRoles(oldRoles, newRoles []string) (toAdd, toRemove []string) {
```
Comment (marcosuma) [inline internal/service/clouduserprojectassignment/model.go:78]: not urgent, can be done in a follow-up ticket - but this method should have its unit test, or you can have a test for `NewAtlasUpdateReq` that checks it always generates the right request based on diff

File: internal/service/clouduserprojectassignment/resource.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+			resp.Diagnostics.AddError(
+				fmt.Sprintf("Error adding role %s to user(%s) in ProjectID(%s):", addReq.GroupRole, username, projectID),
+				err.Error(),
+			)
+			return
+		}
+	}
+
+	for _, removeReq := range removeRequests {
+		_, _, err := connV2.MongoDBCloudUsersApi.RemoveProjectRole(ctx, projectID, userID, removeReq).Execute()
```
Comment (marcosuma) [inline internal/service/clouduserprojectassignment/resource.go:182]: not blocking - if this API fails, the update will be partially applied (only the addprojectRole). This should be documented at minimum, but also we should check if we're consistent in similar situations

cc @maastha

File: internal/service/clouduserprojectassignment/resource.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	ok, parts := conversion.ImportSplit(req.ID, 2)
+	if !ok {
+		resp.Diagnostics.AddError("invalid import ID format", "expected 'project_id/user_id' or 'project_id/username', got: "+importID)
+		return
+	}
+	projectID, user := parts[0], parts[1]
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("project_id"), projectID)...)
+
+	emailRegex := regexp.MustCompile(`^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$`)
```
Comment (marcosuma) [inline internal/service/clouduserprojectassignment/resource.go:245]: not blocking - we don't have any util method that does that already?

File: internal/service/clouduserprojectassignment/model_test.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+func TestNewProjectUserRequest(t *testing.T) {
+	ctx := t.Context()
+	expectedRoles, _ := types.SetValueFrom(ctx, types.StringType, testProjectRoles)
+
+	testCases := map[string]struct {
+		plan     *clouduserprojectassignment.TFModel
+		expected *admin.GroupUserRequest
+	}{
+		"Complete model": {
```
Comment (maastha) [inline internal/service/clouduserprojectassignment/model_test.go:148]: also add test cases for empty/nil values (for all unit tests)

File: internal/service/clouduserprojectassignment/resource.go (hunk 3 above)
Comment (maastha) [reply internal/service/clouduserprojectassignment/resource.go:182]: by documented, do you mean add a comment here or mention something in the public documentation?

File: internal/service/clouduserprojectassignment/model.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	addProjectUserReq := admin.GroupUserRequest{
+		Username: plan.Username.ValueString(),
+		Roles:    roleNames,
+	}
+	return &addProjectUserReq, nil
+}
+
+func NewAtlasUpdateReq(ctx context.Context, plan, state *TFModel) (addRequests, removeRequests []*admin.AddOrRemoveGroupRole, diags diag.Diagnostics) {
```
Comment (maastha) [inline internal/service/clouduserprojectassignment/model.go]: I'm thinking if instead of getting roles from the state here, would it be better to compare against what is returned by the API instead? That ways any partially applied calls can be handled correctly WDYT?

File: internal/service/clouduserprojectassignment/resource.go (hunk 3 above)
Comment (marcosuma) [reply internal/service/clouduserprojectassignment/resource.go:182]: I think we should publicly document it? because I can see a question coming, like: "the update failed but then I logged in and saw some users were added ..."

File: internal/service/clouduserprojectassignment/resource_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/testutil/acc"
+)
+
+var resourceName = "mongodbatlas_cloud_user_project_assignment.test"
+
+func TestAccCloudUserProjectAssignmentRS_basic(t *testing.T) {
+	resource.ParallelTest(t, *basicTestCase(t))
+}
+
+func basicTestCase(t *testing.T) *resource.TestCase {
```
Comment (maastha) [inline internal/service/clouduserprojectassignment/resource_test.go:29]: should we also have a test case to import an existing (ACTIVE) project user? is that possible?

File: internal/service/clouduserprojectassignment/model.go (hunk 6 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/model.go]: Do you mean making another API call before calling this method in the Update?

File: internal/service/clouduserprojectassignment/model.go (hunk 2 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/model.go:78]: Added a test for `NewAtlasUpdateReq`

**Review by jwilliams-mongo — COMMENTED** [review]
@csanx are there any docs updates for this feature? If not, LGTM

**Comment by jwilliams-mongo** [comment]
> @csanx are there any docs updates for this feature? If not, LGTM

I see https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3578/files so I'll review that and approve this

File: internal/service/clouduserprojectassignment/resource_schema.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+			},
+			"first_name": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "First or given name that belongs to the MongoDB Cloud user.",
+			},
+			"project_id": schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "Unique 24-hexadecimal digit string that identifies your project. Use the [/groups](#tag/Projects/operation/listProjects) endpoint to retrieve all projects to which the authenticated user has access.\n\n**NOTE**: Groups and projects are synonymous terms. Your group id is the same as your project id. For existing groups, your group/project id remains the same. The resource and corresponding endpoints use the term groups.",
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(regexp.MustCompile("^([a-f0-9]{24})$"), ""),
```
Comment (AgustinBettati) [inline internal/service/clouduserprojectassignment/resource_schema.go]: Would remove the regex validation just for consistency with other resources

File: internal/service/clouduserprojectassignment/resource.go [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+		resp.Diagnostics.AddError(fmt.Sprintf("error deleting user(%s) from ProjectID(%s):", username, projectID), err.Error())
+		return
+	}
+}
+
+func (r *rs) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	importID := req.ID
+	ok, parts := conversion.ImportSplit(req.ID, 2)
+	if !ok {
+		resp.Diagnostics.AddError("invalid import ID format", "expected 'project_id/user_id' or 'project_id/username', got: "+importID)
```
Comment (AgustinBettati) [inline internal/service/clouduserprojectassignment/resource.go]: q: Why do we support import with `user_id` as well? Does it help with a particular migration journey? Asking as this does add some complexity to our read operation.

File: internal/service/clouduserprojectassignment/resource_migration_test.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+					},
+				},
+				Config: removeProjectInvitationConfigThird(username, projectName, orgID, roles),
+			},
+			mig.TestStepCheckEmptyPlan(removeProjectInvitationConfigThird(username, projectName, orgID, roles)),
+		},
+	})
+}
+
+func originalConfigFirst(username, projectName, orgID string, roles []string) string {
```
Comment (AgustinBettati) [inline internal/service/clouduserprojectassignment/resource_migration_test.go]:
```suggestion
func legacyProjectInvitationConfig(username, projectName, orgID string, roles []string) string {
```
Would make more explicit this is using the deprecated resource

File: internal/service/clouduserprojectassignment/resource_migration_test.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+			project_id  = mongodbatlas_project.mig_test.id
+			username    = local.username
+			roles       = local.roles
+		}
+
+		resource "mongodbatlas_cloud_user_project_assignment" "user_mig_test" {
+			project_id = mongodbatlas_project.mig_test.id
+			username   = local.username
+			roles      = local.roles
+		}
```
Comment (AgustinBettati) [inline internal/service/clouduserprojectassignment/resource_migration_test.go:106]: Can `user_project_assignment` be created with `project_invitation` already existing? I suppose this is how the API behaves but wanted to confirm.

File: internal/service/clouduserprojectassignment/resource_migration_test.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+		resource "mongodbatlas_project" "mig_test" {
+			name 	= %[3]q
+			org_id  = %[4]q
+		}
+
+		resource "mongodbatlas_cloud_user_project_assignment" "user_mig_test" {
+			project_id = mongodbatlas_project.mig_test.id
+			username   = local.username
+			roles      = local.roles
+		}
```
Comment (AgustinBettati) [inline internal/service/clouduserprojectassignment/resource_migration_test.go:127]: When `project_invitation` is deleted it has not impact on `cloud_user_project_assignment`?

**Review by AgustinBettati — COMMENTED** [review]
LGTM, leaving some doubts which seem more API related

File: internal/service/clouduserprojectassignment/resource.go (hunk 4 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource.go:245]: There's a method for this but in the acceptance tests package. Since this same regex is duplicated across 2 of the other new resources, we could consider extracting it into a shared utility function in a follow-up ticket, WDYT?

**Comment by csanx** [comment]
@jwilliams-mongo Yes, my upcoming PRs will add documentation for all new resources and data sources. The one you mention is related to another resource actually

File: internal/service/clouduserprojectassignment/model_test.go (hunk 5 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/model_test.go:148]: Done [59950ba](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3568/commits/59950ba14111760da2db4b341de818fcbc15433e)

File: internal/service/clouduserprojectassignment/resource_migration_test.go (hunk 11 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource_migration_test.go:106]: Yes, after running this, the state shows:
```
{
      "mode": "managed",
      "type": "mongodbatlas_project_invitation",
      "name": "mig_test",
      ...
    },
{
      "mode": "managed",
      "type": "mongodbatlas_cloud_user_project_assignment",
      "name": "user_mig_test",
      ...
    },
```
We can also see in Cloud Dev that the same user is invited twice.

File: internal/service/clouduserprojectassignment/resource_migration_test.go (hunk 12 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource_migration_test.go:127]: No, they act as two independent invitations. Deleting one just removes that specific invitation.

File: internal/service/clouduserprojectassignment/resource.go (hunk 4 above)
Comment (marcosuma) [reply internal/service/clouduserprojectassignment/resource.go:245]: yup, no problem

File: internal/service/clouduserprojectassignment/resource_test.go (hunk 7 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource_test.go:29]: Done

File: internal/service/clouduserprojectassignment/model.go (hunk 6 above)
Comment (maastha) [reply internal/service/clouduserprojectassignment/model.go]: yes

File: internal/service/clouduserprojectassignment/resource.go (hunk 3 above)
Comment (maastha) [reply internal/service/clouduserprojectassignment/resource.go:182]: yes but after failure when they run terraform plan again they will see the exact status of what roles were added vs which were not so I think that should suffice

**Review by marcosuma — APPROVED** [review]

File: internal/service/clouduserprojectassignment/resource.go (hunk 3 above)
Comment (marcosuma) [reply internal/service/clouduserprojectassignment/resource.go:182]: ok, you have a fair point with this one. I'd say an extra clarification in the doc doesn't hurt, especially in case of customers reaching out. But it's a nice to have.

File: internal/service/clouduserprojectassignment/resource_migration_test.go (hunk 11 above)
Comment (AgustinBettati) [reply internal/service/clouduserprojectassignment/resource_migration_test.go:106]: I see, thanks. This sentence for the Tech design helped me understand this migration journey, potentially we can leave it as a comment:

> we only need to consider the migration path for mongodbatlas_project_invitation resources with a PENDING invite (similar to org_invites) - which are not discoverable by the new APIs - so the only way to migrate here is to re-create the pending users.

File: internal/service/clouduserprojectassignment/resource.go (hunk 9 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource.go]: It’s mainly for consistency, since we provide both keys in all Cloud User resources

File: internal/service/clouduserprojectassignment/resource_migration_test.go (hunk 11 above)
Comment (csanx) [reply internal/service/clouduserprojectassignment/resource_migration_test.go:106]: Got it, I’ll take that into account

File: .github/workflows/acceptance-tests-runner.yml [hunk 13]
```diff
@@ -601,8 +602,10 @@ jobs:
           MONGODB_ATLAS_PROJECT_OWNER_ID: ${{ inputs.mongodb_atlas_project_owner_id }}
           MONGODB_ATLAS_ORG_ID: ${{ inputs.mongodb_atlas_org_id }}
           MONGODB_ATLAS_USERNAME: ${{ vars.MONGODB_ATLAS_USERNAME }}
+          MONGODB_ATLAS_USERNAME_2: ${{ vars.MONGODB_ATLAS_USERNAME_2 }}
```
Comment (maastha) [inline .github/workflows/acceptance-tests-runner.yml:605]: is there any difference between MONGODB_ATLAS_USERNAME & MONGODB_ATLAS_USERNAME_2? or we're just adding for more test cases? thinking if we should add a comment here on what MONGODB_ATLAS_USERNAME_2 signifies

**Review by jwilliams-mongo — APPROVED** [review]

**Review by maastha — APPROVED** [review]

File: .github/workflows/acceptance-tests-runner.yml (hunk 13 above)
Comment (csanx) [reply .github/workflows/acceptance-tests-runner.yml:605]: I believe `MONGODB_ATLAS_USERNAME` has the Org Owner role, so if no Project Owner is specified, that user is assigned automatically. Then, when I try to use the `user_project_assignment` resource, I get this error:
`POST: HTTP 400 Bad Request (Error code: "USER_ALREADY_IN_GROUP")`
That’s why I used `MONGODB_ATLAS_USERNAME_2` instead. I’ll add a comment in the code to clarify this.
